java-checks/src/main/java/org/sonar/java/checks/security/HostnameVerifierImplementationCheck.java

```diff
@@ -0,0 +1,126 @@
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S3510")
+public class HostnameVerifierImplementationCheck extends IssuableSubscriptionVisitor {
+
+  private static final String ISSUE_MESSAGE = "Do not unconditionally return true in this method.";
+
+  private static final TypeCriteria TYPE_CRITERIA_STRING = TypeCriteria.is("java.lang.String");
+  private static final TypeCriteria TYPE_CRITERIA_SSL_SESSION = TypeCriteria.is("javax.net.ssl.SSLSession");
+  private static final MethodMatcher VERIFY_METHOD_MATCHER = MethodMatcher.create().typeDefinition(TypeCriteria.subtypeOf("javax.net.ssl.HostnameVerifier"))
+    .name("verify")
+    .addParameter(TYPE_CRITERIA_STRING)
+    .addParameter(TYPE_CRITERIA_SSL_SESSION);
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.kind().equals(Tree.Kind.METHOD)) {
+      checkMethodDefinition((MethodTree) tree);
+    } else if (tree.kind().equals(Tree.Kind.METHOD_INVOCATION)) {
+      ((MethodInvocationTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    } else {
+      ((NewClassTree) tree).arguments().forEach(this::checkForLambdaImplementation);
+    }
+  }
+
+  private void checkMethodDefinition(MethodTree tree) {
+    BlockTree blockTree = tree.block();
+    if (VERIFY_METHOD_MATCHER.matches(tree) && blockTree != null) {
+      checkBlock(blockTree);
+    }
+  }
+
+  private void checkForLambdaImplementation(ExpressionTree argument) {
+    if (argument.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      LambdaExpressionTree lambdaExpressionTree = ((LambdaExpressionTree) argument);
+      Tree lambdaBody = lambdaExpressionTree.body();
+      if (isHostnameVerifierSignature(lambdaExpressionTree)) {
+        if (lambdaBody.is(Tree.Kind.BLOCK)) {
+          checkBlock((BlockTree) lambdaBody);
+        } else if ((lambdaBody.is(Tree.Kind.PARENTHESIZED_EXPRESSION) || lambdaBody.is(Tree.Kind.BOOLEAN_LITERAL))
+          && isTrueLiteral((ExpressionTree) lambdaBody)) {
+          reportIssue(lambdaBody, ISSUE_MESSAGE);
+        }
+      }
+    }
+  }
+
+  private void checkBlock(BlockTree blockTree) {
+    BlockTree innerBlock = blockTree;
+    while (innerBlock.body().size() == 1 && innerBlock.body().get(0).is(Tree.Kind.BLOCK)) {
+      innerBlock = (BlockTree) innerBlock.body().get(0);
+    }
+
+    if (isReturnTrueStatement(innerBlock.body())) {
+      reportIssue(innerBlock.body().get(0), ISSUE_MESSAGE);
+    }
+  }
+
+  private boolean isHostnameVerifierSignature(LambdaExpressionTree lambdaExpressionTree) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Make "isHostnameVerifierSignature" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=1963&issues=AWJr5eRRRMG3VNtqdmei&open=AWJr5eRRRMG3VNtqdmei)

[ISSUE_KEY]: AWJr5eRRRMG3VNtqdmei